wiki/Concept-Starting-and-orchestrating-Bots.md

```diff
@@ -22,17 +25,65 @@ I would recommend an infrastructure that would look like the following diagram:
+
+### Websockets
+
+When using websockets we would open a bidirectional communication channel between the server side and the client side.Thus a great solution because real time communication is possible. Still, it is time consuming to implement and a successful connection is not always garantued as either element (proxy, firewall, …) prevent WebSocket connections between the client and the server or the browser does not support it yet.
+
+### Server-sent events
+
+This is an approach based on the concept of publish-subscribe. The client (desktop app) would subscribe to a 'channel' and the server would push new events (robots to be executed) on that channel. One of the main problem with SSE is the max of 6 client connections from a single host. Also, it is (subjectively) not the best approach to implement with node.js.
+
+### Higher Abstraction Frameworks
+
+There are some frameworks using one (or more) of these basic approaches but are one abstraction layer higher for the user. In particular, I looked at pusher and socket.io.
+Pusher is based on a publish subscribe approach simular to SSE and it seems to be very easy to use. The biggest disadvantage was that it is not open-source but ultimatly one would have to pay for it. Socket.io is a highly popular (> 4 Mio. weekly npm downloads) open source framework. It is mainly based on websocket communication but is also able to seamlessly switch to a long polling approach in the background if websockets do not work.
+
+### Decision
+
+In the end, I decided for socket.io as it is open-cource, well-supported, 'easy' to use, robust and websocket based enabling a bidirectional communication.
+To get started I can recommend reading [this](https://socket.io/docs/v4/index.html) introduction to socket.io. Especially these two subpages [1](https://socket.io/docs/v4/server-socket-instance/) & [2](https://socket.io/docs/v4/client-socket-instance/) are relevant for our usecase.
+
+## How the new approach with socket.io was implemented in the ark automate product
+
+**Desktop app:**
+Here we implemented a CLI that reads the userId the user enters, saves the userId and uses the userId as an authentication for the communication with the web app server. Once started, the desktop app connects with the server by using a socket connection. A socket connection is this bidirectional connection that exist between every client and server. Moreover we use the userId the user entered and ask the server if this socket connection can join the room userId (we have one room for every userId). Once the socket connection was added to the userId room we wait for robots to be executed. Because we are in the userId room we just receive robot execution jobs of web frontends that are connected to the same room.
```

Comment:
passive again

Reply:
```suggestion
Here we implemented a CLI that reads the userId the user enters, saves the userId and uses the userId as an authentication for the communication with the web app server. Once started, the desktop app connects with the server by using a socket connection. A socket connection is this bidirectional connection that exist between every client and server. Moreover we use the userId the user entered and ask the server if this socket connection can join the room userId (we have one room for every userId). Once the socket connection was added to the userId room we wait for robots to be executed. Because we are in the userId room we receive robot execution jobs of web frontends that are connected to the same room.
```